presto-spi/src/main/java/com/facebook/presto/spi/block/BlockEncodingBuffers.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import io.airlift.slice.SliceOutput;
+
+import static java.lang.Math.max;
+
+public abstract class BlockEncodingBuffers
+{
+    protected int[] nextBatchOfPositions;
+    protected int nextBatchSize;
+    protected int bufferedRowCount;
+
+    abstract void prepareBuffers();
+
+    abstract void appendLongValues(long[] values);
+
+    abstract void appendNulls(boolean mayHaveNull, boolean[] nulls);
+
+    abstract void appendOffsets(int[] offsets);
+
+    abstract void appendBlock(Block block);
+
+    abstract void writeTo(SliceOutput sliceOutput);
+
+    abstract int getSizeInBytes();
+
+    public void updateBufferedRowCount(int newRowCount)
+    {
+        bufferedRowCount += newRowCount;
+    }
+
+    public void copyPositions(int[] positions, int startIndex, int batchSize)
+    {
+        if (nextBatchOfPositions == null || nextBatchOfPositions.length < batchSize) {
```

Comment:
extract this logic into a helper method to avoid copy-paste:

```
    private void ensurePositionsCapacity(int length)
    {
        if (nextBatchOfPositions == null || nextBatchOfPositions.length < length) {
            nextBatchOfPositions = new int[max(nextBatchOfPositions.length * 2, length)];
        }
    }
```